src/section/section_proof_chain.rs

```diff
@@ -39,258 +38,203 @@ impl SectionProofBlock {
+    pub(crate) fn push(&mut self, block: SectionProofBlock) {
+        if !validate_next_block(self.last_key_info(), &block) {
+            log_or_panic!(
+                log::Level::Error,
+                "Invalid next block: {:?} -> {:?}",
+                self.last_key_info(),
+                block
+            );
+            return;
+        }
+
+        self.tail.push(block)
     }
 
-    #[cfg(all(test, feature = "mock"))]
-    pub fn all_prefix_version(&self) -> impl DoubleEndedIterator<Item = (&Prefix<XorName>, u64)> {
-        iter::once((&self.prefix, self.version)).chain(
-            self.blocks
-                .iter()
-                .map(|block| (block.prefix(), block.version())),
-        )
+    #[cfg(test)]
+    pub fn self_validate(&self) -> bool {
+        let mut current = &self.head;
+        for block in &self.tail {
+            if !validate_next_block(current, block) {
+                return false;
+            }
+
+            current = block.key_info();
+        }
+        true
     }
 
-    pub fn last_key_info(&self) -> Option<&SectionKeyInfo> {
-        self.blocks.last().map(|block| block.key_info())
+    pub(crate) fn first_key_info(&self) -> &SectionKeyInfo {
+        &self.head
     }
 
-    fn last_trusted_key_info<'a>(
-        &'a self,
-        last_trusted: &'a SectionKeyInfo,
-    ) -> Option<&'a SectionKeyInfo> {
-        let block_offset = last_trusted.version().saturating_sub(self.version) as usize;
+    pub(crate) fn last_key_info(&self) -> &SectionKeyInfo {
+        self.tail
+            .last()
+            .map(|block| block.key_info())
+            .unwrap_or(&self.head)
+    }
 
-        if block_offset == 0 {
-            if last_trusted.version() != self.version || last_trusted.prefix() != &self.prefix {
-                return None;
-            }
-        } else if let Some(block) = self.blocks.get(block_offset - 1) {
-            if block.key_info() != last_trusted {
-                return None;
-            }
-        } else {
-            // Root of trust not found
-            return None;
+    #[cfg(all(test, feature = "mock"))]
+    pub fn key_infos(&self) -> impl DoubleEndedIterator<Item = &SectionKeyInfo> {
+        iter::once(&self.head).chain(self.tail.iter().map(|block| block.key_info()))
+    }
+
+    /// Returns a slice of this chain starting at the given index.
+    pub(crate) fn slice_from(&self, first_index: usize) -> Self {
+        if first_index == 0 || self.tail.is_empty() {
+            return self.clone();
         }
 
-        let mut current = last_trusted;
-        for block in &self.blocks[block_offset..] {
-            if !validate_next_block(current, block) {
-                return None;
-            }
+        let head_index = std::cmp::min(first_index, self.tail.len()) - 1;
+        let head = self.tail[head_index].key_info().clone();
+        let tail = self.tail[head_index + 1..].iter().cloned().collect();
```

Comment:
say there are only two blocks (one in head, one in tail) currently, and the first_index to slice_from is 1.
Then, `head_index` is 0. so wondering `self.tail[head_index + 1..].iter()` will cause a panic or just an empty vec ?

Reply:
Yeah, I also needed to check this, but turns our it returns empty vec, allowing us to keep the code simple: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b5c886758865c4be0b1eff24c90980aa

Reply:
good to know that :)